config/file.go

```diff
@@ -0,0 +1,323 @@
+	environmentOverrides map[string]interface{}
+	configLock           sync.RWMutex
+	path                 string
+	watch                bool
+	watcher              *watcher
+	needsSave            bool
+}
+
+// NewFileStore creates a new instance of a config store backed by the given file path.
+//
+// If watch is true, any external changes to the file will force a reload.
+func NewFileStore(path string, watch bool) (fs *fileStore, needsSave bool, err error) {
+	resolvedPath, err := resolveConfigFilePath(path)
+	if err != nil {
+		return nil, false, err
+	}
+
+	fs = &fileStore{
+		path:  resolvedPath,
+		watch: watch,
+	}
+	if _, err = fs.Load(); err != nil {
+		return nil, false, err
+	}
+
+	if fs.watch {
+		if err = fs.startWatcher(); err != nil {
+			mlog.Error("failed to start config watcher", mlog.String("path", path), mlog.Err(err))
+		}
+	}
+
+	return fs, fs.needsSave, nil
+}
+
+// resolveConfigFilePath attempts to resolve the given configuration file path to an absolute path.
+//
+// Consideration is given to maintaining backwards compatibility when resolving the path to the
+// configuration file.
+func resolveConfigFilePath(path string) (string, error) {
+	// Absolute paths are explicit and require no resolution.
+	if filepath.IsAbs(path) {
+		return path, nil
+	}
+
+	// Search for the given relative path (or plain filename) in various directories,
+	// resolving to the corresponding absolute path if found. FindConfigFile takes into account
+	// various common search paths rooted both at the current working directory and relative
+	// to the executable.
+	if configFile := fileutils.FindConfigFile(path); configFile != "" {
+		return configFile, nil
+	}
+
+	// Otherwise, search for the config/ folder using the same heuristics as above, and build
+	// an absolute path anchored there and joining the given input path (or plain filename).
+	if configFolder, found := fileutils.FindDir("config"); found {
+		return filepath.Join(configFolder, path), nil
+	}
+
+	// Fail altogether if we can't even find the config/ folder. This should only happen if
+	// the executable is relocated away from the supporting files.
+	return "", fmt.Errorf("failed to find config file %s", path)
+}
+
+// Get fetches the current, cached configuration.
+func (fs *fileStore) Get() *model.Config {
+	fs.configLock.RLock()
+	defer fs.configLock.RUnlock()
+
+	return fs.config
+}
+
+// GetEnvironmentOverrides fetches the configuration fields overridden by environment variables.
+func (fs *fileStore) GetEnvironmentOverrides() map[string]interface{} {
+	fs.configLock.RLock()
+	defer fs.configLock.RUnlock()
+
+	return fs.environmentOverrides
+}
+
+// Set replaces the current configuration in its entirety, without updating the backing store.
+func (fs *fileStore) Set(newCfg *model.Config) (*model.Config, error) {
+	fs.configLock.Lock()
+	var unlockOnce sync.Once
+	defer func() {
+		unlockOnce.Do(func() {
+			fs.configLock.Unlock()
+		})
+	}()
+
+	oldCfg := fs.config
```

Comment:
Concurrently, "bad" code could still be updating the `fs.config` returned by `Get` above. I'd like to force a `Clone` in `Get` as discussed, but deferred for now.